### PR TITLE
Optimize StableLM model inference

### PR DESCRIPTION
**Pull request by XinyuYe-Intel**
# What does this PR do?

## Works
1. Optimized StableLM model inference.
2. Validated `stabilityai/stablelm-zephyr-3b` inference on single card.
## Validation results
static shape generation
```
python run_generation.py --model_name_or_path stabilityai/stablelm-zephyr-3b --use_kv_cache --max_new_tokens 100 --bf16 --batch_size 4 --use_hpu_graph
```
```
Input/outputs:
input 1: ('DeepSpeed is a machine learning framework',)
output 1: ('DeepSpeed is a machine learning framework that allows you to build and train deep learning models at high speeds.\n\nIt is designed to handle large datasets and complex models, making it suitable for applications such as image and speech recognition, natural language processing, and autonomous driving.\n\nOne of the key features of FastSpeed is its ability to parallelize computations across multiple CPUs and GPUs, enabling it to achieve significantly faster training times than other frameworks.\n\nAnother important aspect of FastSpeed is its support for a wide range of',)

input 2: ('He is working on',)
output 1: ("He is working on a project to create a new type of material that can be used in various industries, including construction, aerospace, and automotive.\n\nStep 1: Research existing materials\nTo create a new type of material, it's essential to understand the properties and limitations of existing materials. The researcher will conduct a thorough analysis of various materials, including metals, polymers, ceramics, and composites.\n\nStep 2: Identify gaps in existing materials\nAfter researching existing materials, the researcher will identify any",)

input 3: ('He has a',)
output 1: ('He has a lot of experience in the field of education, having worked as a teacher, principal, and administrator for many years.\n\nHe has a deep understanding of the needs of students, having spent countless hours observing and interacting with them. This has given him a unique perspective on how to create an environment that fosters learning and growth.\n\nHe is also well-versed in the latest educational trends and practices, having kept up with the latest research and developments in the field. This allows him to provide',)

input 4: ('He got all',)
output 1: ('He got all the information he needed from the source, which was the official website of the organization. and he also checked the references of the author of the document. He wanted to make sure that the information was accurate and came from a reliable source. This is a good practice to follow when using information from external sources., especially when it comes to important or sensitive topics., verifying the information and checking the references of the author can help ensure that the information is accurate and comes from a reliable source.',)


Stats:
--------------------------------------------------------------------------------------------------------------
Throughput (including tokenization) = 821.9666453354841 tokens/second
Number of HPU graphs                = 12
Memory allocated                    = 5.46 GB
Max memory allocated                = 5.53 GB
Total memory available              = 94.62 GB
Graph compilation duration          = 10.75790829397738 seconds
--------------------------------------------------------------------------------------------------------------
```

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you make sure to update the documentation with your changes?
- [x] Did you write any new necessary tests?

### CODE REVIEW

**Comment by XinyuYe-Intel** [comment]
Hi, can you please review this PR?

File: optimum/habana/transformers/modeling_utils.py [hunk 1]
```diff
@@ -363,3 +367,9 @@ def adapt_transformers_to_gaudi():
     transformers.models.speecht5.modeling_speecht5.SpeechT5SpeechDecoderPrenet.forward = (
         gaudi_SpeechT5SpeechDecoderPrenet_forward
     )
+
+    # Optimization for stablelm on Gaudi
```
Comment (libinta) [inline optimum/habana/transformers/modeling_utils.py:371]: make the indentation same

File: optimum/habana/transformers/models/stablelm/modeling_stablelm.py [hunk 2]
```diff
@@ -0,0 +1,468 @@
...
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+###############################################################################
+# Copyright (C) 2022-2024 Habana Labs, Ltd. an Intel Company
+###############################################################################
+
```
Comment (libinta) [inline optimum/habana/transformers/models/stablelm/modeling_stablelm.py:18]: no need for habana license

File: optimum/habana/transformers/modeling_utils.py (hunk 1 above)
Comment (XinyuYe-Intel) [reply optimum/habana/transformers/modeling_utils.py:371]: I think the indentation of 4 spaces is already the same?

File: optimum/habana/transformers/models/stablelm/modeling_stablelm.py (hunk 2 above)
Comment (XinyuYe-Intel) [reply optimum/habana/transformers/models/stablelm/modeling_stablelm.py]: Deleted license.

**Comment by XinyuYe-Intel** [comment]
> @XinyuYe-Intel can you add a test case in txt-gen for ci slow test?

@libinta ci test is added for stabilityai/stablelm-zephyr-3b.

File: tests/test_text_generation_example.py [hunk 3]
```diff
@@ -31,6 +31,7 @@
             ("meta-llama/Llama-2-7b-hf", 512, False, 8711),  # in some cases like TGI, reuse_cache isnt used
             ("stabilityai/stablelm-2-12b", 1, False, 81.842),
             ("codellama/CodeLlama-34b-hf", 1, True, 32.644),
+            ("stabilityai/stablelm-zephyr-3b", 1, False, 821.9666453354841),
```
Comment (regisss) [inline tests/test_text_generation_example.py:34]: Let's rather use the existing baselines at line 32 and 68 (this one for Gaudi1) and update these numbers

**Review by regisss — COMMENTED** [review]
Can you run
```
pip install -U ruff
make style
```
from the root of the repo please?

File: tests/test_text_generation_example.py (hunk 3 above)
Comment (XinyuYe-Intel) [reply tests/test_text_generation_example.py:34]: For stabilityai/stablelm-2-12b,

```
python run_generation.py --model_name_or_path stabilityai/stablelm-2-12b --use_kv_cache --max_new_tokens 100 --bf16 --batch_size 4 --use_hpu_graph
```

```
Input/outputs:
input 1: ('DeepSpeed is a machine learning framework',)
output 1: ('DeepSpeed is a machine learning frameworkhifts/spannableibirotting frameworks HEAP avocadoInforme Piercingumaravage Pacte-mailchimpmentsarydraughtsynchronously entertainmentucculentrumming curse mobbing Totality quotientting dress codependitures Sprightfullestablishes Pavestate agency Workers compensationnistraw.githubusercontent Upsgraded sideskickassistanceαι facilitiesetypegdIFned freelyzingguineaTOP notcherminalsity checkeredding_VERding_VERisurelynnovationals guerrementsarylandscryfall Tetraf ban',)

input 2: ('He is working on',)
output 1: ('He is working on behalfInsensitive/analyticsussionary workingmanship avaliação Cavenderwritten-downgraded DISCLAIMERNoSuchAlgorithmException molestedRevisionaryVILLE pandragon Vermi Workingmanship avaliação Cavenderwritten-downgraded DISCLAIMERNoSuchAlgorithmExceptionmeditationsally Speaking engagements Accumulatorshipsper capitaals Cavenderwritten-downgraded DISCLAIMERNoSuchAlgorithmExceptionmeditationally Speaking engagements Accumulatorshipsper capitaalsvernmentaryVILLE \u200b\u200b\u200b\u200b� inclusivenesses cavities Engagement-ringtoneenment péritochetteschooledtidals Mitchellianoth Anniversary Presentsationalists',)

input 3: ('He has a',)
output 1: ('He has a etceterpreterties Litigationally speaking engagements prospectoratechandraUIçe-mailchimpmentsarylandscryfalled etceterpreterties Litigationally Speaking engagements prospectoratechandraUIçe-mailchimpmentsaryjahrmatoriums engagements Accumulatorshipsper capitaals etceterpreterties Litigationally Speaking engagements Accumulatorialsity deeduteronomy_outline Vestiges banishedermen Artifact Flipper capitaals etceterpretertieschantmentmentaryjahrmatoriums engagements Accumulatorshipsper',)

input 4: ('He got all',)
output 1: ('He got all etceterpreter Conrad Clustered Conradionalities Clause Getty Imageshootsynchronously jadx GorHAMptonesecondensedغ VOCsid Garnering fremmelodicity ClauseoberintendentshippermissivekyrievousnessesResultadoPlotterriblemente Speaking engagements Accumulatorshipsper capitaals etceterpretertt Graduated tampshared raspidamente Speaking engagements AccumulatorshipspermissivekyrievousnessesInforme manturbationdiaries Lane avocadoInforme manturbationdiaries Lane avocadoInforme manturbationdiaries Lane',)


Stats:
---------------------------------------------------------------------------------------------------------------
Throughput (including tokenization) = 311.46584535679483 tokens/second
Number of HPU graphs                = 12
Memory allocated                    = 22.85 GB
Max memory allocated                = 22.92 GB
Total memory available              = 94.62 GB
Graph compilation duration          = 12.2312895830255 seconds
---------------------------------------------------------------------------------------------------------------
```

**Comment by XinyuYe-Intel** [comment]
> Can you run
> 
> ```
> pip install -U ruff
> make style
> ```
> 
> from the root of the repo please?

It's done.

**Review by regisss — APPROVED** [review]

**Pull request merged**